File: graph-coloring/GraphColoring.java
import java.util.Scanner;

public class GraphColoring {
    static int n;
    static int m;
    static boolean[][] graph;
    static int[] colors;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the number of vertices: ");
        n = scanner.nextInt();
        System.out.print("Enter the number of colors: ");
        m = scanner.nextInt();

        graph = new boolean[n][n];
        System.out.println("Enter the adjacency matrix of the graph (0 for false, 1 for true):");
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                graph[i][j] = scanner.nextInt() == 1;
            }
        }

        colors = new int[n];

        if (mColoring(0)) {
            for (int i = 0; i < n; i++) {
                System.out.println("Vertex " + (i + 1) + ": Color " + colors[i]);
            }
        } else {
            System.out.println("Graph cannot be colored with " + m + " colors.");
        }

        scanner.close();
    }

    static boolean mColoring(int k) {
        if (k == n) {
            return true;
        }

        for (int color = 1; color <= m; color++) {
            if (isLegalColor(k, color)) {
                colors[k] = color;
                if (mColoring(k + 1)) {
                    return true;
                }
                colors[k] = 0; // Backtrack
            }
        }
        return false;
    }

    static boolean isLegalColor(int vertex, int color) {
        for (int i = 0; i < n; i++) {
            if (graph[vertex][i] && colors[i] == color) {
                return false;
            }
        }
        return true;
    }
}

File: graph-coloring/graph-coloring.cpp
#include <iostream>
#include <vector>

using namespace std;

bool isSafe(vector<vector<int>>& graph, vector<int>& color, int v, int c) {
    for (int i = 0; i < graph[v].size(); i++) {
        if (graph[v][i] == 1 && color[i] == c) {
            return false;
        }
    }
    return true;
}

bool graphColoringUtil(vector<vector<int>>& graph, int m, vector<int>& color, int v) {
    if (v == graph.size()) {
        for (int i = 0; i < graph.size(); i++) {
            cout << color[i] << " ";
        }
        cout << endl;
        return true; 
    }
    for (int c = 1; c <= m; c++) {
        if (isSafe(graph, color, v, c)) {
            color[v] = c;

            if (graphColoringUtil(graph, m, color, v + 1)) {
                // Found a valid coloring, continue searching
            }

            color[v] = 0; 
        }
    }
    return false;
}

void graphColoring(vector<vector<int>>& graph, int m) {
    int V = graph.size();
    vector<int> color(V, 0);

    graphColoringUtil(graph, m, color, 0);
}

int main() {
    int V;
    cout << "Enter the number of vertices: ";
    cin >> V;

    vector<vector<int>> graph(V, vector<int>(V, 0)); 

    cout << "Enter the adjacency matrix:" << endl;
    for (int i = 0; i < V; i++) {
        for (int j = 0; j < V; j++) {
            cin >> graph[i][j];
        }
    }

    int m; 
    cout << "Enter the number of colors: ";
    cin >> m;

    graphColoring(graph, m);
    return 0;
}
